Mathlib/MeasureTheory/FoelnerFilter.lean

```diff
@@ -0,0 +1,138 @@
+/-
+Copyright (c) 2025 Stefano Rocca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Stefano Rocca
+-/
+import Mathlib.MeasureTheory.Group.Defs
+import Mathlib.MeasureTheory.Group.Action
+import Mathlib.MeasureTheory.Measure.Typeclasses.Finite
+
+/-!
+# Følner filters - definitions and properties
+
+This file defines Følner filters for measurable spaces acted on by a group.
+
+## Definitions
+
+* `IsFoelner G μ l` : A Følner filter on a measurable `G`-space `X` with measure `μ`
+  (a measurable space with an action of `G`) is a filter `l` on `Set X` such that:
+    1. Each `s in l` is eventually measurable with finite non-zero measure,
+    2. For all `g : G`, `μ (g • s ∆ s) / μ (s)` tends to `0`.
+
+* `MaximalFoelner G μ` : The maximal Følner filter for a measurable `G`-space `X`
+  with measure `μ` is the pullback of `𝓝 0` along the map `s ↦ μ (g • s) / μ s` on measurable
+  sets of finite non-zero measure.
+
+-/
+
+open MeasureTheory Filter Set
+open scoped ENNReal Pointwise symmDiff Topology Filter
+
+variable {G X : Type*} [MeasurableSpace X] [Group G] [MulAction G X] {μ : Measure X}
+
+namespace Filter
+
+variable (G μ) in
+/-- A Følner filter on a measurable `G`-space `X` with measure `μ`
+(a measurable space with an action of `G`) is a filter `l` on `Set X` such that:
+    1. Each `s in l` is eventually measurable with finite non-zero measure,
+    2. For all `g : G`, `μ (g • s ∆ s) / μ (s)` tends to `0`. -/
+structure IsFoelner (l : Filter (Set X)) : Prop where
+  meas_set : ∀ᶠ s in l, MeasurableSet s
+  meas_ne_zero : ∀ᶠ s in l, μ s ≠ 0
+  meas_ne_top : ∀ᶠ s in l, μ s ≠ ∞
+  tendsto_meas_symmDiff (g : G) : Tendsto (fun s ↦ μ ((g • s) ∆ s) / μ s) l (𝓝 0)
+
+/-- The constant filter `X` is Følner if `X` has finite measure. -/
+lemma IsFoelner.pure_of_isFiniteMeasure [NeZero μ] [IsFiniteMeasure μ] :
+    IsFoelner G μ (pure .univ) where
+  meas_set := by simp
+  meas_ne_zero := by simp [NeZero.ne]
+  meas_ne_top := by simp
+  tendsto_meas_symmDiff (g : G) := by
+    simpa using tendsto_pure_nhds (fun s ↦ μ ((g • s) ∆ s) / μ s) .univ
+
+/-- If there exists a non-trivial Følner filter on the measurable `G`-space `X`,
+then it exists a `G`-invariant finitely additive probability measure on `X`. -/
+lemma IsFoelner.amenable [SMulInvariantMeasure G X μ] {l : Filter (Set X)} [NeBot l]
+    (hfoel : IsFoelner G μ l) : ∃ m : Set X → ℝ≥0∞, m .univ = 1 ∧
+    (∀ s t, MeasurableSet t → Disjoint s t → m (s ∪ t) = m s + m t) ∧
+    ∀ (g : G) (s : Set X), m (g • s) = m s := by
```

Comment:
Avoid starting multiline return types in the middle of a line. It is very unreadable
```suggestion
    (hfoel : IsFoelner G μ l) :
    ∃ m : Set X → ℝ≥0∞, m .univ = 1 ∧
      (∀ s t, MeasurableSet t → Disjoint s t → m (s ∪ t) = m s + m t) ∧
      ∀ (g : G) (s : Set X), m (g • s) = m s := by
```